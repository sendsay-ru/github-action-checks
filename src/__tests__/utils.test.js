const { render } = require('../utils');
const source = require('../../config/source.json');
const emptyConfig = require('../../config/empty');
const defaultConfig = require('../../config/default');
const sendsasyConfig = require('../../config/sendsay');

describe('Render', () => {
  it('should expose a function', () => {
    expect(render).toBeDefined();
  });

  it('should return expected value with empty config', () => {
    expect(render(source, emptyConfig)).toMatchSnapshot();
  });

  it('should return expected value with default config', () => {
    expect(render(source, defaultConfig)).toMatchSnapshot();
  });

  it('should return expected value with sendsay config', () => {
    expect(render(source, sendsasyConfig)).toMatchSnapshot();
  });
});
