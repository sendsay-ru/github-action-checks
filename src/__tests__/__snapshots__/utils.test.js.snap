// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render should return expected value with default config 1`] = `
[
  {
    "content": "name: Check PR
on: [pull_request]

env:
  NODE_ENV: CI

concurrency:
  group: \${{ github.workflow }}-\${{ github.event.pull_request.number }}
  cancel-in-progress: true

jobs:

  checks:

    runs-on: ubuntu-latest

    outputs:
      eslint: \${{ steps.eslint.outputs.result }}
      prettier: \${{ steps.prettier.outputs.result }}

    steps:
      - name: Checkout
        uses: actions/checkout@v3

      - name: Preparing
        uses: ./.github/actions/check-install
        with:
          cache-key: cache-turbo-checks-\${{ github.event.pull_request.number }}

      - name: Eslint
        id: eslint
        if: always()
        uses: ./.github/actions/check-run
        with:
          script: npm eslint

      - name: Prettier
        id: prettier
        if: always()
        uses: ./.github/actions/check-run
        with:
          script: npm pretify

  tests:

    runs-on: ubuntu-latest

    outputs:
      unit: \${{ steps.unit.outputs.result }}

    steps:
      - name: Checkout
        uses: actions/checkout@v3

      - name: Preparing
        uses: ./.github/actions/check-install
        with:
          cache-key: cache-turbo-checks-\${{ github.event.pull_request.number }}

      - name: Unit
        id: unit
        if: always()
        uses: ./.github/actions/check-run
        with:
          script: npm test


",
    "output": ".github/workflows/check-pr.yml",
  },
  {
    "content": "name: Check production
on:
  push:
    branches:
      - stable
      - master
    tags:
      - v**

env:
  NODE_ENV: CI
  FORCE_COLOR: 1

jobs:

  checks:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout
        uses: actions/checkout@v3

      - name: Preparing
        uses: ./.github/actions/check-install
        with:
          cache-key: cache-turbo-checks-\${{ github.event.ref }}

      - name: Eslint
        id: eslint
        if: always()
        uses: ./.github/actions/check-run
        with:
          script: npm eslint

      - name: Prettier
        id: prettier
        if: always()
        uses: ./.github/actions/check-run
        with:
          script: npm pretify

  tests:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout
        uses: actions/checkout@v3

      - name: Preparing
        uses: ./.github/actions/check-install
        with:
          cache-key: cache-turbo-checks-\${{ github.event.ref }}

      - name: Unit
        id: unit
        if: always()
        uses: ./.github/actions/check-run
        with:
          script: npm test

",
    "output": ".github/workflows/check-prod.yml",
  },
  {
    "content": "name: Check comment
description: show comment with results

inputs:
  token:
    required: true
    description: GITHUB_TOKEN
  title:
    required: true
    description: title
  eslint:
    required: true
    description: Eslint
  prettier:
    required: true
    description: Prettier
  unit:
    required: true
    description: Unit

runs:
  using: composite
  steps:
      - name: Comment
        uses: sendsay-ru/rewritable-pr-comment@v1.0.1
        with:
          message: "\${{ inputs.title }} \\n\\n
                  | Eslint | Prettier | Unit |\\n
                  | --- | --- | --- |\\n
                  |  \${{ inputs.eslint }}  |  \${{ inputs.prettier }}  |  \${{ inputs.unit }}  |\\n\\n
                  [üîó See details](\${{ github.server_url }}/\${{ github.repository }}/actions/runs/\${{ github.run_id }})"
          GITHUB_TOKEN: \${{ inputs.token }}
          COMMENT_IDENTIFIER: "comment-check-result"
",
    "output": ".github/actions/check-comment/action.yml",
  },
  {
    "content": "name: Preparing
description: preparing to run the checks

inputs:
  cache-key:
    required: true
    description: build cache key

runs:
  using: composite
  steps:
    - name: Setup node
      uses: actions/setup-node@v3
      with:
        node-version: 16

    - name: Install packages
      shell: bash
      run: npm install

",
    "output": ".github/actions/check-install/action.yml",
  },
  {
    "content": "name: Run check
description: run check script

inputs:
  name:
    required: true
    description: name
  script:
    required: true
    description: script command

outputs:
  result:
    description: "Script result"
    value: \${{ steps.result.outputs.status }}

runs:
  using: composite
  steps:
    - name: Run script
      id: script
      shell: bash
      run: \${{ inputs.script }}

    - name: Results
      id: result
      if: always()
      env:
        SUCCESS_ICON: "‚òòÔ∏è"
        FAILURE_ICON: "üêô"
      shell: bash
      run: |
        if [[ \${{ steps.script.outcome }} == 'success' ]]; then echo "status=$SUCCESS_ICON" >> $GITHUB_OUTPUT; else echo "status=$FAILURE_ICON" >> $GITHUB_OUTPUT; fi
",
    "output": ".github/actions/check-run/action.yml",
  },
]
`;

exports[`Render should return expected value with empty config 1`] = `
[
  {
    "content": "name: Check PR
on: [pull_request]

env:
  NODE_ENV: CI

concurrency:
  group: \${{ github.workflow }}-\${{ github.event.pull_request.number }}
  cancel-in-progress: true

jobs:


",
    "output": ".github/workflows/check-pr.yml",
  },
  {
    "content": "name: Check production
on:
  push:
    branches:
      - stable
      - master
    tags:
      - v**

env:
  NODE_ENV: CI
  FORCE_COLOR: 1

jobs:

",
    "output": ".github/workflows/check-prod.yml",
  },
  {
    "content": "name: Check comment
description: show comment with results

inputs:
  token:
    required: true
    description: GITHUB_TOKEN
  title:
    required: true
    description: title

runs:
  using: composite
  steps:
      - name: Comment
        uses: sendsay-ru/rewritable-pr-comment@v1.0.1
        with:
          message: "\${{ inputs.title }} \\n\\n
                  |  |\\n
                  |  |\\n
                  |  |\\n\\n
                  [üîó See details](\${{ github.server_url }}/\${{ github.repository }}/actions/runs/\${{ github.run_id }})"
          GITHUB_TOKEN: \${{ inputs.token }}
          COMMENT_IDENTIFIER: "comment-check-result"
",
    "output": ".github/actions/check-comment/action.yml",
  },
  {
    "content": "name: Preparing
description: preparing to run the checks

inputs:
  cache-key:
    required: true
    description: build cache key

runs:
  using: composite
  steps:
    - name: Setup node
      uses: actions/setup-node@v3
      with:
        node-version: 16

",
    "output": ".github/actions/check-install/action.yml",
  },
  {
    "content": "name: Run check
description: run check script

inputs:
  name:
    required: true
    description: name
  script:
    required: true
    description: script command

outputs:
  result:
    description: "Script result"
    value: \${{ steps.result.outputs.status }}

runs:
  using: composite
  steps:
    - name: Run script
      id: script
      shell: bash
      run: \${{ inputs.script }}

    - name: Results
      id: result
      if: always()
      env:
        SUCCESS_ICON: "‚òòÔ∏è"
        FAILURE_ICON: "üêô"
      shell: bash
      run: |
        if [[ \${{ steps.script.outcome }} == 'success' ]]; then echo "status=$SUCCESS_ICON" >> $GITHUB_OUTPUT; else echo "status=$FAILURE_ICON" >> $GITHUB_OUTPUT; fi
",
    "output": ".github/actions/check-run/action.yml",
  },
]
`;

exports[`Render should return expected value with sendsay config 1`] = `
[
  {
    "content": "name: Check PR
on: [pull_request]

env:
  NODE_ENV: CI

concurrency:
  group: \${{ github.workflow }}-\${{ github.event.pull_request.number }}
  cancel-in-progress: true

jobs:
  start:
    runs-on: ubuntu-latest
    env:
      FORCE_COLOR: 1
      PROGRESS_TITLE: "### üë∑ Checks in progress..."
      PROGRESS_ICON: "üèÑüèª‚Äç‚ôÇÔ∏è"

    steps:
      - name: Checkout
        uses: actions/checkout@v3

      - name: Starting comment
        uses: ./.github/actions/check-comment
        with:
          title: \${{ env.PROGRESS_TITLE }}
          prettier: \${{ env.PROGRESS_ICON }}
          stylelint: \${{ env.PROGRESS_ICON }}
          typescript: \${{ env.PROGRESS_ICON }}
          eslint: \${{ env.PROGRESS_ICON }}
          unit: \${{ env.PROGRESS_ICON }}
          unit: \${{ env.PROGRESS_ICON }}
          token: \${{ secrets.GITHUB_TOKEN }}

  checks:
    needs: start
    runs-on: ubuntu-latest

    outputs:
      prettier: \${{ steps.prettier.outputs.result }}
      stylelint: \${{ steps.stylelint.outputs.result }}
      typescript: \${{ steps.typescript.outputs.result }}

    steps:
      - name: Checkout
        uses: actions/checkout@v3

      - name: Preparing
        uses: ./.github/actions/check-install
        with:
          cache-key: cache-turbo-checks-\${{ github.event.pull_request.number }}

      - name: Prettier
        id: prettier
        if: always()
        uses: ./.github/actions/check-run
        with:
          script: yarn pretify

      - name: Stylelint
        id: stylelint
        if: always()
        uses: ./.github/actions/check-run
        with:
          script: yarn stylelint

      - name: Typescript
        id: typescript
        if: always()
        uses: ./.github/actions/check-run
        with:
          script: yarn dev-tsc

  eslint:
    needs: start
    runs-on: ubuntu-latest

    outputs:
      eslint: \${{ steps.eslint.outputs.result }}

    steps:
      - name: Checkout
        uses: actions/checkout@v3

      - name: Preparing
        uses: ./.github/actions/check-install
        with:
          cache-key: cache-turbo-checks-\${{ github.event.pull_request.number }}

      - name: Eslint
        id: eslint
        if: always()
        uses: ./.github/actions/check-run
        with:
          script: yarn eslint

  unit-app:
    needs: start
    runs-on: ubuntu-latest

    outputs:
      unit: \${{ steps.unit.outputs.result }}

    steps:
      - name: Checkout
        uses: actions/checkout@v3

      - name: Preparing
        uses: ./.github/actions/check-install
        with:
          cache-key: cache-turbo-checks-\${{ github.event.pull_request.number }}

      - name: Unit-app
        id: unit
        if: always()
        uses: ./.github/actions/check-run
        with:
          script: yarn unit-app

  unit-rest:
    needs: start
    runs-on: ubuntu-latest

    outputs:
      unit: \${{ steps.unit.outputs.result }}

    steps:
      - name: Checkout
        uses: actions/checkout@v3

      - name: Preparing
        uses: ./.github/actions/check-install
        with:
          cache-key: cache-turbo-checks-\${{ github.event.pull_request.number }}

      - name: Unit-rest
        id: unit
        if: always()
        uses: ./.github/actions/check-run
        with:
          script: yarn unit-packages


  report:
    needs: [checks,eslint,unit-app,unit-rest]
    if: always()
    runs-on: ubuntu-latest
    env:
      SUCCESS_TITLE: "### ‚úÖ Checks completed successfully!"
      FAILURE_TITLE:  "### üö´ Checks were executed with errors!"
    steps:
      - name: Checkout
        uses: actions/checkout@v3

      - name: Title
        id: title
        run: |
          if [[ \${{ needs.checks.result }} == 'success' ]] && [[ \${{ needs.eslint.result }} == 'success' ]] && [[ \${{ needs.unit-app.result }} == 'success' ]] && [[ \${{ needs.unit-rest.result }} == 'success' ]]; then echo "str=\${{ env.SUCCESS_TITLE }}" >> $GITHUB_OUTPUT; else echo "str=\${{ env.FAILURE_TITLE }}" >> $GITHUB_OUTPUT; fi

      - name: Final comment
        uses: ./.github/actions/check-comment
        with:
          title: \${{ steps.title.outputs.str }}
          prettier: \${{ needs.checks.outputs.prettier }}
          stylelint: \${{ needs.checks.outputs.stylelint }}
          typescript: \${{ needs.checks.outputs.typescript }}
          eslint: \${{ needs.eslint.outputs.eslint }}
          unit: \${{ needs.unit-app.outputs.unit }}
          unit: \${{ needs.unit-rest.outputs.unit }}
          token: \${{ secrets.GITHUB_TOKEN }}

",
    "output": ".github/workflows/check-pr.yml",
  },
  {
    "content": "name: Check production
on:
  push:
    branches:
      - stable
      - master
    tags:
      - v**

env:
  NODE_ENV: CI
  FORCE_COLOR: 1

jobs:

  checks:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout
        uses: actions/checkout@v3

      - name: Preparing
        uses: ./.github/actions/check-install
        with:
          cache-key: cache-turbo-checks-\${{ github.event.ref }}

      - name: Prettier
        id: prettier
        if: always()
        uses: ./.github/actions/check-run
        with:
          script: yarn pretify

      - name: Stylelint
        id: stylelint
        if: always()
        uses: ./.github/actions/check-run
        with:
          script: yarn stylelint

      - name: Typescript
        id: typescript
        if: always()
        uses: ./.github/actions/check-run
        with:
          script: yarn dev-tsc

  eslint:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout
        uses: actions/checkout@v3

      - name: Preparing
        uses: ./.github/actions/check-install
        with:
          cache-key: cache-turbo-checks-\${{ github.event.ref }}

      - name: Eslint
        id: eslint
        if: always()
        uses: ./.github/actions/check-run
        with:
          script: yarn eslint

  unit-app:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout
        uses: actions/checkout@v3

      - name: Preparing
        uses: ./.github/actions/check-install
        with:
          cache-key: cache-turbo-checks-\${{ github.event.ref }}

      - name: Unit-app
        id: unit
        if: always()
        uses: ./.github/actions/check-run
        with:
          script: yarn unit-app

  unit-rest:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout
        uses: actions/checkout@v3

      - name: Preparing
        uses: ./.github/actions/check-install
        with:
          cache-key: cache-turbo-checks-\${{ github.event.ref }}

      - name: Unit-rest
        id: unit
        if: always()
        uses: ./.github/actions/check-run
        with:
          script: yarn unit-packages

",
    "output": ".github/workflows/check-prod.yml",
  },
  {
    "content": "name: Check comment
description: show comment with results

inputs:
  token:
    required: true
    description: GITHUB_TOKEN
  title:
    required: true
    description: title
  prettier:
    required: true
    description: Prettier
  stylelint:
    required: true
    description: Stylelint
  typescript:
    required: true
    description: Typescript
  eslint:
    required: true
    description: Eslint
  unit:
    required: true
    description: Unit-app
  unit:
    required: true
    description: Unit-rest

runs:
  using: composite
  steps:
      - name: Comment
        uses: sendsay-ru/rewritable-pr-comment@v1.0.1
        with:
          message: "\${{ inputs.title }} \\n\\n
                  | Prettier | Stylelint | Typescript | Eslint | Unit-app | Unit-rest |\\n
                  | --- | --- | --- | --- | --- | --- |\\n
                  |  \${{ inputs.prettier }}  |  \${{ inputs.stylelint }}  |  \${{ inputs.typescript }}  |  \${{ inputs.eslint }}  |  \${{ inputs.unit }}  |  \${{ inputs.unit }}  |\\n\\n
                  [üîó See details](\${{ github.server_url }}/\${{ github.repository }}/actions/runs/\${{ github.run_id }})"
          GITHUB_TOKEN: \${{ inputs.token }}
          COMMENT_IDENTIFIER: "comment-check-result"
",
    "output": ".github/actions/check-comment/action.yml",
  },
  {
    "content": "name: Preparing
description: preparing to run the checks

inputs:
  cache-key:
    required: true
    description: build cache key

runs:
  using: composite
  steps:
    - name: Setup node
      uses: actions/setup-node@v3
      with:
        node-version: 16

    - name: Install packages
      shell: bash
      run: yarn install

    - name: Build packages
      if: always()
      shell: bash
      run: yarn build-packages

",
    "output": ".github/actions/check-install/action.yml",
  },
  {
    "content": "name: Run check
description: run check script

inputs:
  name:
    required: true
    description: name
  script:
    required: true
    description: script command

outputs:
  result:
    description: "Script result"
    value: \${{ steps.result.outputs.status }}

runs:
  using: composite
  steps:
    - name: Run script
      id: script
      shell: bash
      run: \${{ inputs.script }}

    - name: Results
      id: result
      if: always()
      env:
        SUCCESS_ICON: "‚òòÔ∏è"
        FAILURE_ICON: "üêô"
      shell: bash
      run: |
        if [[ \${{ steps.script.outcome }} == 'success' ]]; then echo "status=$SUCCESS_ICON" >> $GITHUB_OUTPUT; else echo "status=$FAILURE_ICON" >> $GITHUB_OUTPUT; fi
",
    "output": ".github/actions/check-run/action.yml",
  },
]
`;
